// @vibrant/core/dist/cjs/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const options = require("./options.cjs");
const builder = require("./builder.cjs");
const utils = require("./utils.cjs");
const index = require("./pipeline/index.cjs");
const client = require("./pipeline/worker/client.cjs");
const host = require("./pipeline/worker/host.cjs");
const _Vibrant = class _Vibrant {
  constructor(_src, opts) {
    this._src = _src;
    this.opts = utils.assignDeep({}, _Vibrant.DefaultOpts, opts);
  }
  static use(pipeline) {
    this._pipeline = pipeline;
  }
  static from(src) {
    return new builder.Builder(src);
  }
  get result() {
    return this._result;
  }
  _process(image, opts) {
    image.scaleDown(this.opts);
    const processOpts = options.buildProcessOptions(this.opts, opts);
    return _Vibrant._pipeline.process(image.getImageData(), processOpts);
  }
  palette() {
    return this.swatches();
  }
  swatches() {
    throw new Error(
      "Method deprecated. Use `Vibrant.result.palettes[name]` instead"
    );
  }
  async getPalette() {
    const arg0 = arguments[0];
    const arg1 = arguments[1];
    const name = typeof arg0 === "string" ? arg0 : "default";
    const cb = typeof arg0 === "string" ? arg1 : arg0;
    const image = new this.opts.ImageClass();
    try {
      const image1 = await image.load(this._src);
      const result1 = await this._process(image1, {
        generators: [name]
      });
      this._result = result1;
      const res = result1.palettes[name];
      if (!res) {
        throw new Error(`Palette with name ${name} not found`);
      }
      image.remove();
      if (cb) {
        cb(void 0, res);
      }
      return res;
    } catch (err) {
      image.remove();
      if (cb) {
        cb(err);
      }
      return Promise.reject(err);
    }
  }
  async getPalettes() {
    const arg0 = arguments[0];
    const arg1 = arguments[1];
    const names = Array.isArray(arg0) ? arg0 : ["*"];
    const cb = Array.isArray(arg0) ? arg1 : arg0;
    const image = new this.opts.ImageClass();
    try {
      const image1 = await image.load(this._src);
      const result1 = await this._process(image1, {
        generators: names
      });
      this._result = result1;
      const res = result1.palettes;
      image.remove();
      if (cb) {
        cb(void 0, res);
      }
      return res;
    } catch (err) {
      image.remove();
      if (cb) {
        cb(err);
      }
      return Promise.reject(err);
    }
  }
};
_Vibrant.DefaultOpts = {
  colorCount: 64,
  quality: 5,
  filters: []
};
let Vibrant = _Vibrant;
exports.Builder = builder.Builder;
exports.BasicPipeline = index.BasicPipeline;
exports.WorkerPipeline = client.WorkerPipeline;
exports.runPipelineInWorker = host.runPipelineInWorker;
exports.Vibrant = Vibrant;
